File: codigos/tads/linear/pilha.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    int *elem;
    int cont;
} Pilha;

Pilha *newPilha(int size) {
    Pilha *pilha;
    pilha = (Pilha *)malloc(sizeof(Pilha));
    pilha->elem = (int *)malloc(size * sizeof(int));
    pilha->cont = 0;

    return pilha;
}

void delPilha(Pilha *pilha) {
    free(pilha);
}

void inserir(Pilha *pilha, int x) {
    if (pilha->cont >= sizeof(pilha)) return;

    pilha->elem[pilha->cont] = x;
    pilha->cont++;
}

int remover(Pilha *pilha) {
    if (pilha->cont == 0) return pilha->cont;

    return pilha->elem[--pilha->cont];
}

void mostrar(Pilha *pilha) {
    for (int i = 0; i < pilha->cont; i++) {
        printf("%d ", pilha->elem[i]);
    }
    printf("\n");
}

int main() {
    Pilha *pilha = newPilha(6);

    inserir(pilha, 3);
    inserir(pilha, 5);
    inserir(pilha, 8);
    inserir(pilha, 9);
    inserir(pilha, 7);
    inserir(pilha, 2);
    mostrar(pilha);

    remover(pilha);
    remover(pilha);
    remover(pilha);
    mostrar(pilha);

    return 0;
}

File: codigos/tads/linear/Pilha.java
class Pilha {
    public static void main(String[] args) {
        CriarPilha pilha = new CriarPilha(6);

        pilha.empilhar(5);
        pilha.empilhar(4);
        pilha.empilhar(2);
        pilha.empilhar(13);
        pilha.empilhar(8);
        pilha.mostrar();

        pilha.remover();
        pilha.remover();
        pilha.remover();
        pilha.mostrar();
    }    
}

class CriarPilha {
    int[] array;
    int cont;

    CriarPilha(int size) {
        array = new int[size];
        cont = 0;
    }

    void empilhar(int x) {
        if (cont >= array.length) {
            System.err.println();
            return;
        }

        array[cont] = x;
        cont++;
    }

    int remover() {
        if (cont == 0) {
            System.err.println();
            return cont;
        }

        return array[--cont];
    }

    void mostrar() {
        for (int i = 0; i < cont; i++) {
            System.out.print(array[i] + " ");
        }
        System.out.println();
    }
}
